New graph component

MCLiii opened the pull request.

jrvollmer commented on Frontend/src/Components/Dashboard/Dashboard.js:
```diff
@@ -104,34 +77,9 @@ async function callBackendSingleValuesAPI() {
 
   const [state, setState] = useState({ data: null });
-  useEffect(() => {
+  const updateData = async () => {
```
Why are we doing this? I thought it was for graph data. What if we instead move this API call into a child component of Dashboard that houses the left-hand side of the screen (FaultsView and the 4 panes in the lower-left) so that re-rendering quickly isn't an issue? The graphs wouldn't be directly affected by that state change since it's no longer in the parent

MCLiii replied:
Okay

jrvollmer commented on Frontend/src/Components/Dashboard/Dashboard.js:
```diff
@@ -4,7 +4,7 @@ import {
   HStack,
   Select,
-  useColorMode
+  useColorMode, useInterval
```
Note that this would need to be removed also per previous comment

MCLiii replied:
好

jrvollmer commented on Frontend/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "chase-car-dashboard-frontend",
-  "version": "4.3.0",
+  "version": "5.0.0",
```
I don't want to be _too_ nit-picky, but the frontend version bump from this and the data recording changes should be changed so that either:
- We wait to bump the frontend version at all until the python backend refactor is good
- We change the data recording bump to be a major one on both the backend and frontend sides, and this will be a major bump as well. Both changes implement new api calls in the backend and frontend, so it would be a non-backwards compatible change on both ends since they're versioned separately.

I think the first option will be less of a hassle. Just delete the "bump versions" commit that Jack made and remove this change from your "Getting graph to work" commit. Then, hold off on any version updates until `python-backend` is good to merge. Once it is, do a major version bump on all sections of the dashboard.

jrvollmer replied:
Reverted on new-graph-component. Wait until python-backend is good to rebase onto main before moving version bump to the end of the commit history.

jrvollmer commented on Frontend/src/Components/Graph/CustomGraph.js:
```diff
@@ -63,18 +59,23 @@ for (const category of GraphData.Output) {
   // console.log("now:", now);
   // console.log("diff:", DateTime.now().diff(now).toHuman());
 
+  /**
+ * Stores all the units in the graph-data json file as key: value pairs
+ */
+
```
Move back to original position

jrvollmer commented on Frontend/src/Components/Graph/CustomGraph.js:
```diff
@@ -63,18 +59,23 @@ for (const category of GraphData.Output) {
   }
 }
 
+
 // the options fed into the graph object, save regardless of datasets
```
```suggestion

// the options fed into the graph object, save regardless of datasets
```
Fixing unnecessary whitespace change

jrvollmer replied:
Addressed

jrvollmer commented on Frontend/src/Components/Graph/CustomGraph.js:
```diff
@@ -63,18 +59,23 @@ for (const category of GraphData.Output) {
   return {
     responsive: true,
-    maintainAspectRatio: false,
+    maintainAspectRatio: true,
```
```suggestion
    maintainAspectRatio: false,
```
This is probably your issue with the graphs not fitting vertically

jrvollmer commented on Frontend/src/Components/Graph/CustomGraph.js:
```diff
@@ -93,28 +94,32 @@ function getOptions(now, secondsRetained, colorMode) {
+              units[item.dataset.label]
             }`;
           },
+          labelColor: (item) => {
+            return {
+              backgroundColor: genColor(item.dataset.label),
+            }
+          },
```
Tooltip background color?

MCLiii replied:
Correct

jrvollmer commented on Frontend/src/Components/Graph/CustomGraph.js:
```diff
@@ -306,50 +287,24 @@ export default function CustomGraph(props) {
-  // but the graph has no name associated with it
+  }, [isDataSelectOpen, onDataSelectClose, datasetKeys, historyLength, title, onSave]);
+
+
```
Can we just remove the problematic line and keep the original formatting?

jrvollmer commented on Frontend/src/Components/Graph/CustomGraph.js:
```diff
@@ -159,139 +162,117 @@ function getOptions(now, secondsRetained, colorMode) {
+ * @param props
+ * @constructor
+ */
+function Graph(props) {
```
I could be wrong, but I have a sneaking suspicion that this could be defined within the `CustomGraph` component to help reduce the electric boogaloo of things being redefined

jrvollmer commented on Frontend/src/Components/Graph/CustomGraph.js:
```diff
@@ -306,50 +287,24 @@ export default function CustomGraph(props) {
     ),
-    [isNameModalOpen, onNameModalClose, onNewSave]
+    [isNameModalOpen, onNameModalClose]
   );
 
-  // the actual graph shown to the user
-  const graph = useMemo(() => {
-    // const min = DateTime.now().minus(historyLength * 1_000);
-    console.time("graph update");
-    const temp = (
-      <Line
-        data={{ datasets: formattedDatasets }}
-        options={getOptions(
-          DateTime.fromFormat(latestTime, "HH:mm:ss.SSS") ?? DateTime.now(),
-          historyLength,
-          colorMode
-        )}
-        parsing="false"
-      />
-    );
-    console.timeEnd("graph update");
-    return temp;
-  }, [formattedDatasets, historyLength, latestTime]);
 
-  return (
+  return(
     <>
-      <HStack w="100%" h="100%" align="stretch" {...stackProps}>
+      <HStack w="100%" h="100%" align="stretch">
```
Why'd we remove this? This might be another issue causing the graph vertical overflow

jrvollmer commented on Frontend/src/Components/Graph/CustomGraph.js:
```diff
@@ -390,52 +345,44 @@ export default function CustomGraph(props) {
               flex="1 1 0"
               spacing={2}
               borderWidth={1}
-              borderColor={borderCol}
```
This is for dark mode. Let's revert it

jrvollmer commented on Frontend/src/Components/Graph/CustomGraph.js:
```diff
@@ -390,52 +345,44 @@ export default function CustomGraph(props) {
-}
+    </>);
+};
+
```
Revert to minimize changes

jrvollmer reviewed (changes requested):
@MCLiii, let's discuss the graph changes during the work session. I also want to fix the history of the `python-backend` branch and this one, but I can take care of that. I just want to run it by you before any new changes are made here

jrvollmer commented on package-lock.json:
Shouldn't need chart dependencies in top-level package.json

jrvollmer commented:
@MCLiii @jrvollmer, note that using the Docker RedisTimeSeries image runs into issues with running the graphs for more than a minute. Running the module locally rather than with Docker works.

jrvollmer commented on Backend/components/graph_api.py:
```diff
@@ -6,9 +6,15 @@
     """pass in a comma separated list of data that the frontend requests start_time end_time must be in unix millisec"""
+    if not data:
+        # use a query to get timestamp
+        query_result = await db.query(['speed'], start_time, end_time, ['avg'],
+                                      (end_time - start_time) // 60 or 1)
+        return {'response': {"timestamps": query_result.index.to_list()}}
     data = [d.strip() for d in data.split(',')]
-    query_result = await db.query(data, start_time, end_time, ['avg']*len(data), (end_time-start_time)//5000 or 1)
-    result = {"timestamp": query_result.index.to_list()}
+    query_result = await db.query(data, start_time, end_time, ['avg']*len(data), (end_time-start_time)//60 or 1)
+    result = {"timestamps": query_result.index.to_list()}
```
We need to use min-max decimation for booleans, and we should ensure that any outstanding/non-nominal values are shown on the graph. For reference, this is a boolean value:
![image](https://github.com/badgerloop-software/chase-car-dashboard/assets/72111610/294c1c3f-f7a1-4bb4-b0ba-95d81316d013)

MCLiii replied:
Yeah, I'm thinking adding a new field to dataformat to specify the aggregation algorithm for the data

jrvollmer replied:
Good idea

jrvollmer replied:
@MCLiii, once this is added (or at least when we can use a min-max algorithm for booleans and an average for other values), the PR should pretty much be ready to go

MCLiii replied:
Boolean aggregation is fixed

jrvollmer replied:
Thanks!

jrvollmer reviewed (changes requested):
I'm currently working on some state management and styling fixes

MCLiii reviewed (commented):
Need to fix tool tip not properly displaying for data with different sampling density
<img width="890" alt="Screenshot 2023-12-05 at 11 08 49" src="https://github.com/badgerloop-software/chase-car-dashboard/assets/57700505/3fb1dc81-6db7-4e04-8327-2aa5c23078d2">

jrvollmer commented on Frontend/src/Components/Graph/CustomGraph.js:
```diff
@@ -395,47 +403,44 @@ export default function CustomGraph(props) {
-          <Center flex={1}>{graph}</Center>
+          <Center flex={1}>
+            <Graph
+                querylist={datasetKeys.filter(key=>!noShowKeys[key])}
+                histLen={historyLength}
+                colorMode={colorMode}
+                optionInfo={optionInfo}
+                extremes={yRange}
+              />
+            </Center>
```
```suggestion
              querylist={datasetKeys.filter(key=>!noShowKeys[key])}
              histLen={historyLength}
              colorMode={colorMode}
              optionInfo={optionInfo}
              extremes={yRange}
            />
          </Center>
```

jrvollmer commented on package-lock.json:
Need to update versions in multiple places in package-lock.json. After changing the version in `package.json`, you can run `npm i` to update the versions in `/**/package-lock.json`

jrvollmer commented on Frontend/src/Components/Graph/CustomGraph.js:
```diff
@@ -34,12 +36,14 @@ import {
 import { DateTime } from "luxon";
-import { useCallback, useEffect, useMemo, useState } from "react";
+import { useEffect, useMemo, useState } from "react";
 import { Line } from "react-chartjs-2";
-import { FaSave } from "react-icons/fa";
+import { FaSave } from "react-icons/all";
```
?

jrvollmer replied:
Reverted

jrvollmer commented on package.json:
A few things:
- There is no backend version update, even though the backend changed.
- The frontend version update was a major version update, so the overall version bump would be a major one as well.
- **All that said, I think that we should not update any versions for this PR since it's being added to the python-refactor branch instead of main.**

jrvollmer replied:
Reverted version bump

jrvollmer reviewed (changes requested):
@MCLiii

jrvollmer commented on Frontend/src/Components/Graph/CustomGraph.js:
```diff
@@ -68,13 +70,47 @@ for (const category of GraphData.Output) {
 }
+ChartJS.register(
+
+)
```
We already registered above

jrvollmer commented on Frontend/src/Components/Graph/CustomGraph.js:
```diff
@@ -50,7 +52,7 @@ ChartJS.register(
   LineElement,
   Title,
   Tooltip,
-  Legend
+  Legend,
```
Remove

jrvollmer commented on Backend/components/graph_api.py:
```diff
@@ -1,14 +1,49 @@
     for d in data:
-        result[d] = query_result[d].to_list()
-    return result
+        # TODO change this to actual aggregation
+        group['no-aggr'].append(d) if config.FORMAT[d]['type']=='bool' else group['avg'].append(d)
```
@MCLiii, remove or address TODO

MCLiii replied:
This will need to wait until signal spreadsheet is updated

jrvollmer replied:
Gotcha, sounds good

jrvollmer commented on Backend/core/comms.py:
```diff
@@ -22,7 +22,8 @@ def gen_format_str(file_path: str):
         properties.append(key)
-    print(byte_length)
+        config.FORMAT[key] = {'type': data_format[key][1]}
+    print(config.FORMAT)
```
Remove print

jrvollmer replied:
done

MCLiii committed:
Graph backend

Add decimation algos:
- Avg for analog values
- Min-max for bools
- TODO Specify algo for each value in data format

MCLiii committed:
Update and improve graph frontend

- Use new graph API and adjust graph-related components accordingly
- Reduce graphSelectModal rerender frequency by removing unnecessary
  dependencies from its useMemo()
- Handle fetch errors/timeouts with new graph API
- Add unix timestamps to faux-queue and misc-constants

Co-authored-by: James <jamesvollmer02@gmail.com>

jrvollmer committed:
Ignore Backend/*.egg-info

jrvollmer reviewed (approved):
LGTM
